sympy/ntheory/qs.py

```diff
@@ -0,0 +1,489 @@
+        self.a = a
+        self.b = b
+
+    def eval(self, x):
+        """
+        Compute the value of the sieve polynomial at point x.
+
+        Parameters:
+        ===========
+
+        x : Integer parameter for sieve polynomial
+        """
+        ans = 0
+        for coeff in self.modified_coeff:
+            ans *= x
+            ans += coeff
+        return ans
+
+class factor_base_elem:
+    """This class stores an element of the factor_base.
+    """
+    def __init__(self, prime, tmem_p, log_p):
+        """
+        Initialization of factor_base_elem.
+
+        Parameters:
+        ===========
+
+        prime : prime number of the factor_base
+        tmem_p : Integer square root of x**2 = n mod prime
+        log_p : Compute Natural Logarithm of the prime
+        """
+        self.prime = prime
+        self.tmem_p = tmem_p
+        self.log_p = log_p
+        self.soln1 = None
+        self.soln2 = None
+        self.a_inv = None
+        self.b_ainv = None
+
+
+def _generate_factor_base(num_primes, n):
+    """Generate factor_base for Quadratic Sieve. The factor_base
+    consists of all the the points whose legendre_symbol(n, p) == 1
+    and p < num_primes. Along with the prime factor_basealso stores
+    natural logarithm of prime and the residue n modulo p.
+
+    Parameters:
+    ===========
+
+    num_primes : upper bound of the factor_base
+    n : integer to be factored
+    """
+    from sympy import prime
+    factor_base = []
+    i = 1
+    while len(factor_base) < num_primes:
+        ith_prime = prime(i)
+        i += 1
+        if pow(n, (ith_prime - 1) // 2, ith_prime) == 1:
+            residue = _sqrt_mod_prime_power(n, ith_prime, 1)[0]
+            log_p = round(log(ith_prime)*2**10)
+            factor_base.append(factor_base_elem(ith_prime, residue, log_p))
+    return factor_base
+
+def _initialize_first_polynomial(N, M, factor_base):
+    """This step is the initialization of the 1st sieve polynomial.
+    Here `a` is selected as a product of several primes of the factor_base
+    such that `a` is about to sqrt(2*N) / M. Other initial values of
+    factor_base elem are also intialized which includes a_inv, b_ainv, soln1,
+    soln2 which are used when the sieve polynomial is changed. The b_ainv
+    is required for fast polynomial change as we don't have to calculate
+    `2*b*mod_inverse(a, prime)` every time.
+
+    Parameters:
+    ===========
+
+    N : Number to be factored
+    M : sieve interval
+    factor_base : factor_base primes
+    """
+    approx_val = sqrt(2*N) / M
+    # `a` is a parameter of the sieve polynomial and `q` is the prime factors of `a`
+    # randomly search for a combination of primes whose multiplication is close to approx_val
+    # This multiplication of primes will be `a` and the primes will be `q`
+    # `best_a` denotes that `a` is close to approx_val in the random search of combination
+    best_a, best_q, best_ratio = None, None, None
+    for _ in range(50):
```

Comment:
Any specific reason to loop it to `50`?

Reply:
There is no reason for `50`. Here we are trying to              select `a` as a product of a subset of primes in factor_base such that it is close to `sqrt(sqrt(2*N) / M)`. But there is no straight forward algorithm to do it, so i am selecting `a` 50 times randomly and finally choosing `a` with          least ratio with `sqrt(2*N) / M`. This 50 times is not necessary        and it can be reduced.

Reply:
> This 50 times is not necessary and it can be reduced.

I see, so how far below we can go without affecting the accuracy? and then the reason for the choice of the number needs to be added as a comment.

Reply:
this step is   inexpensive, reducing 50 to 30 will not have much effect.